lib/src/annotate.rs

```diff
@@ -0,0 +1,335 @@
+        }
+        let remaining_lines = remaining_lines.unwrap();
+        for (_, original_line_number) in remaining_lines {
+            self.original_line_map
+                .insert(original_line_number, commit_id.clone());
+        }
+    }
+
+    fn get_next_commit(&self) -> Option<CommitId> {
+        return self.local_line_map.keys().next().cloned();
+    }
+
+    fn convert_to_results(self, original_contents: &[u8]) -> AnnotateResults {
+        let original_content_lines: Vec<String> =
+            original_contents.lines().map(|s| s.unwrap()).collect();
+        let mut result_lines = Vec::new();
+        for (idx, line) in original_content_lines.into_iter().enumerate() {
+            result_lines.push((self.original_line_map.get(&idx).unwrap().clone(), line));
+        }
+        AnnotateResults {
+            file_annotations: result_lines,
+        }
+    }
+}
+
+/// Get line by line annotations for a specific file path in the repo.
+pub fn get_annotation_for_file(
+    file_name: &str,
+    repo: &ReadonlyRepo,
+    starting_commit_id: &CommitId,
+) -> Result<AnnotateResults, AnnotateError> {
+    let file_path = RepoPathBuf::from_relative_path(file_name);
+    if file_path.is_err() {
+        return Err(AnnotateError::FileNotFound(file_name.to_string()));
+    }
+    let file_path: RepoPathBuf = file_path.unwrap();
+    let store = repo.store();
+
+    let current_commit = store.get_commit(starting_commit_id)?;
+    let current_tree = current_commit.tree()?;
+    let original_file_id = get_file_id(current_commit.id(), &current_tree, &file_path)?;
+    if original_file_id.is_none() {
+        return Err(AnnotateError::FileNotFound(file_name.to_string()));
+    }
+    let original_contents =
+        get_file_contents(repo.store(), &file_path, &original_file_id.unwrap())?;
+    let num_lines = original_contents.lines().count();
+    let mut partial_results = PartialResults::new(starting_commit_id, num_lines);
+
+    process_commits(repo, current_commit.id(), &file_path, &mut partial_results)?;
+
+    Ok(partial_results.convert_to_results(&original_contents))
+}
+
+fn process_commits(
+    repo: &ReadonlyRepo,
+    starting_commit_id: &CommitId,
+    file_name: &RepoPath,
+    results: &mut PartialResults,
+) -> Result<(), AnnotateError> {
+    let mut current_commit = Some(repo.store().get_commit(starting_commit_id)?);
+    while let Some(next_commit) = current_commit {
+        process_commit(&next_commit, repo, file_name, results)?;
+        let next_commit_id = results.get_next_commit();
+        match next_commit_id {
+            None => {
+                current_commit = None;
+            }
+            Some(next_id) => {
+                current_commit = Some(repo.store().get_commit(&next_id)?);
+            }
+        }
+    }
+    Ok(())
+}
+
+fn process_commit(
+    commit: &Commit,
+    repo: &ReadonlyRepo,
+    file_name: &RepoPath,
+    results: &mut PartialResults,
+) -> Result<(), AnnotateError> {
```

Comment:
Ditto

Reply:
For process_commits I can probably return the PartialResults. However, from process_commit, I'd rather pass in a mutable reference that mutates since I call it within a loop. To return it every time means that each call to process_commit needs to consume the old results and return new results which, if not boxed, just wastes unneeded copying on the stack. Therefore, I think process_commit is more ergonomic to pass a mutable reference.

Reply:
fair enough